### PR TITLE
[FLINK-4239] Set Default Allowed Lateness to Zero and Make Triggers N…

### PR DESCRIPTION
**Pull request by aljoscha**
R: @kl0u for review

### CODE REVIEW

**Comment by uce** [comment]
@kl0u can you review this today? This should be part of the 1.1 release.

**Comment by kl0u** [comment]
@uce I will do it today.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/PurgingTrigger.java [hunk 1]
```diff
@@ -44,39 +44,36 @@ private  PurgingTrigger(Trigger<T, W> nestedTrigger) {
...
-			case FIRE:
-				return TriggerResult.FIRE_AND_PURGE;
-			case FIRE_AND_PURGE:
-				return TriggerResult.FIRE_AND_PURGE;
-			default:
-				return TriggerResult.CONTINUE;
+		if (triggerResult.isFire()) {
+			return TriggerResult.FIRE_AND_PURGE;
+		} else if (triggerResult.isPurge()) {
+			return TriggerResult.PURGE;
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/PurgingTrigger.java]: Couldn't we simplify it even more like the following:
`return triggerResult.isFire() ? TriggerResult.FIRE_AND_PURGE : triggerResult;`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/PurgingTrigger.java [hunk 2]
```diff
@@ -44,39 +44,36 @@ private  PurgingTrigger(Trigger<T, W> nestedTrigger) {
...
+		} else {
+			return TriggerResult.CONTINUE;
 		}
 	}
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/PurgingTrigger.java]: Same as above.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/PurgingTrigger.java [hunk 3]
```diff
@@ -44,39 +44,36 @@ private  PurgingTrigger(Trigger<T, W> nestedTrigger) {
...
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/PurgingTrigger.java]: Same as above.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/PurgingTrigger.java [hunk 4]
```diff
@@ -93,13 +90,12 @@ public boolean canMerge() {
...
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/PurgingTrigger.java]: Same as above.

**Comment by aljoscha** [comment]
Thanks for the review! The change you proposed really makes the code a lot nicer. 😃

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java [hunk 5]
```diff
@@ -1409,7 +1409,7 @@ public void testDropDueToLatenessSliding() throws Exception {
 	}
 
 	@Test
-	public void testDropDueToLatenessSessionZeroLateness() throws Exception {
+	public void testDropDueToLatenessSessionZeroLatenessPurgingTrigger() throws Exception {
 		final int GAP_SIZE = 3;
 		final long LATENESS = 0;
```
Comment (kl0u) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java]: Here the name changes to indicate that it is purging, but the Trigger is the same (EventTime) without purging.

File: flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java [hunk 6]
```diff
@@ -114,8 +115,8 @@ private void runTest(
 			windowedStream = windowedStream.allowedLateness(Time.milliseconds(ALLOWED_LATENESS_MS));
 		}
 
-		if (!PURGE_WINDOW_ON_FIRE) {
-			windowedStream = windowedStream.trigger(new NonPurgingEventTimeTriggerWrapper());
+		if (PURGE_WINDOW_ON_FIRE) {
+			windowedStream = windowedStream.trigger(PurgingTrigger.of(EventTimeTrigger.create()));
 		}
```
Comment (kl0u) [inline flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java]: I do not know if this behavior existed before, but now if you set the PURGE_WINDOW_ON_FIRE to true, now the test fails.

**Comment by kl0u** [comment]
Thanks a lot for the work @aljoscha 
I just had some minor comments!

File: flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java (hunk 6 above)
Comment (aljoscha) [inline flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java]: Yes, I think this is on purpose, for now. @StefanRRichter should have more input on that.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java (hunk 5 above)
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java]: Fixing

**Comment by aljoscha** [comment]
I addressed the other comments as well.

File: flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java (hunk 6 above)
Comment (kl0u) [inline flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java]: Thanks @aljoscha

**Comment by kl0u** [comment]
LGTM

File: flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java (hunk 6 above)
Comment (StefanRRichter) [inline flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java]: Previously, the correct behavior in the case was still under discussion and not clearly defined. This is why this branch of the test could not yet be properly implemented and should fail as notification that it needs to be adapted once the final behavior is defined.

**Comment by uce** [comment]
I'm merging this now

**Pull request closed without merging**